### PR TITLE
Feat: upgrade asset system

### PR DESCRIPTION
**Pull request by SupremoUGH**
Third part of #153. It consists of the changes related to the asset system. To be merged after #282.

Changes:
`manta-util`:
- `codec.rs`: `Decode` implementation for `u128`.

`manta-accounting`: 
- `asset.rs`: Upgrade to the new asset system. Now `Asset` is more general and depends on two generics.
- `transfer`: Included the associated types `AssetId` and `AssetValue` in the `Configuration` trait. 
- rest of `manta-accounting`: Minimal changes so it compiles.

`manta-pay`:
- `config`: Concrete implementations of `AssetId` and `AssetValue` for `Config`, with the relevant trait implementations.
- rest of `manta-pay`: minimal changes so it compiles.
---

Before we can merge this PR, please make sure that all the following items have been checked off:

- [x] Linked to an issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** line describing your change in [`CHANGELOG.md`](https://github.com/manta-network/manta-rs/blob/main/CHANGELOG.md) and added the appropriate `changelog` label to the PR.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [x] Checked that changes and commits conform to the standards outlined in [`CONTRIBUTING.md`](https://github.com/manta-network/manta-rs/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by GhostOfGauss — APPROVED** [review]

**Comment by GhostOfGauss** [comment]
@bhgomes do you know why the CI Lint check is skipping?

**Comment by bhgomes** [comment]
> @bhgomes do you know why the CI Lint check is skipping?

Not really sure why, will have to investigate. Probably it's timing out or overusing resources.

File: manta-pay/src/simulation/ledger/mod.rs [hunk 1]
```diff
@@ -452,9 +484,18 @@ impl ledger::Write<Vec<TransferPost>> for LedgerConnection {
...
+impl PublicBalanceOracle<Config> for LedgerConnection {
     #[inline]
-    fn public_balances(&self) -> LocalBoxFuture<Option<AssetList>> {
+    fn public_balances(
+        &self,
+    ) -> LocalBoxFuture<
+        Option<
+            AssetList<
+                <Config as transfer::Configuration>::AssetId,
+                <Config as transfer::Configuration>::AssetValue,
```
Comment (bhgomes) [inline manta-pay/src/simulation/ledger/mod.rs:495]: For all of these kinds of things just make an alias in the `crate::config` file for `AssetId` and `AssetValue` and call them instead.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-accounting/src/asset.rs [hunk 2]
```diff
@@ -54,497 +69,318 @@ use std::{
...
-            self.value -= asset.value;
-            true
-        } else {
-            false
-        }
-    }
 }
 
-impl<I, V> Add<V> for Asset<I, V>
+impl<I, V, COM> eclair::cmp::PartialEq<Self, COM> for Asset<I, V>
```
[the remark this answers is not in the document]
Comment (bhgomes) [reply manta-accounting/src/asset.rs:226]: It's currently in flux right now as the ECLAIR derive model is not stabilized.

File: manta-accounting/src/asset.rs [hunk 3]
```diff
@@ -671,87 +549,110 @@ impl AssetList {
...
+{
     #[inline]
-    fn from_iter<I>(iter: I) -> Self
+    fn from_iter<A>(iter: A) -> Self
     where
-        I: IntoIterator<Item = (AssetId, AssetValue)>,
+        A: IntoIterator<Item = (I, V)>,
     {
         iter.into_iter()
             .map(move |(id, value)| Asset::new(id, value))
```
[the remark this answers is not in the document]
Comment (bhgomes) [reply manta-accounting/src/asset.rs:612]: Yes if `collect` is called, they are sorted.

File: manta-accounting/src/asset.rs [hunk 4]
```diff
@@ -671,87 +549,110 @@ impl AssetList {
...
+    V: AddAssign + Default + PartialEq,
+{
     #[inline]
-    fn from_iter<I>(iter: I) -> Self
+    fn from_iter<A>(iter: A) -> Self
     where
-        I: IntoIterator<Item = AssetList>,
+        A: IntoIterator<Item = AssetList<I, V>>,
     {
         iter.into_iter().map::<Vec<_>, _>(Into::into).collect()
```
[the remark this answers is not in the document]
Comment (bhgomes) [reply manta-accounting/src/asset.rs:643]: same

File: manta-accounting/src/asset.rs [hunk 5]
```diff
@@ -855,78 +759,81 @@ pub trait AssetMap: Default {
...
+                        .any(move |a| &a.id == id && a.value == Default::default())
                         .then(move || key.clone())
                 })
                 .take(n)
                 .collect()
         }
 
         #[inline]
-        fn insert(&mut self, key: Self::Key, asset: Asset) {
+        fn insert(&mut self, key: Self::Key, asset: Asset<$I, $V>) {
```
[the remark this answers is not in the document]
Comment (bhgomes) [reply manta-accounting/src/asset.rs:836]: The keys are meant to be unique in the honest system, if you get duplicate keys you can at least handle them in some cases but you can't necessarily handle them in all cases. This is not a protocol error, we are just being defensive here.

**Review by bhgomes — APPROVED** [review]

**Review by GhostOfGauss — APPROVED** [review]

**Pull request merged**